RUN_TIME/tmp/clisp-2.45/src/clisp.1
.\"     Title: clisp
.\"    Author: Bruno Haible <\fI\%http://www.haible.de/bruno/\fR>
.\" Generator: DocBook XSL Stylesheets vsnapshot_7553 <http://docbook.sf.net/>
.\"      Date: Last modified: 2008-05-12
.\"    Manual: Platform: x86_64-unknown-linux-gnu
.\"    Source: CLISP 2.45
.\"
.TH "CLISP" "1" "Last modified: 2008\-05\-12" "CLISP 2.45" "Platform: x86_64-unknown-linux-gnu"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
clisp - ANSI Common Lisp compiler, interpreter and debugger.
.SH "SYNOPSIS"
.HP 6
\fBclisp\fR [[\fB\-h\fR] | [\fB\-\-help\fR]] [\fB\-\-version\fR] [\fB\-\-license\fR] [\fB\-help\-image\fR] [\fB\-B\ \fIlisp\-lib\-dir\fR\fR] [\fB\-K\ \fIlinking\-set\fR\fR] [\fB\-M\ \fImem\-file\fR\fR] [\fB\-m\ \fImemory\-size\fR\fR]
[\fB\-L\ \fIlanguage\fR\fR] [\fB\-N\ \fIlocale\-dir\fR\fR] [\fB\-E\fIdomain\fR\ \fIencoding\fR\fR] [[\fB\-q\fR] | [\fB\-\-quiet\fR] | [\fB\-\-silent\fR] | [\fB\-v\fR] | [\fB\-\-verbose\fR]] [\fB\-on\-error\ \fIaction\fR\fR] [\fB\-repl\fR] [\fB\-w\fR] [\fB\-I\fR] [[\fB\-ansi\fR] | [\fB\-traditional\fR]] [\fB\-modern\fR] [\fB\-p\ \fIpackage\fR\fR] [\fB\-C\fR] [\fB\-norc\fR] [\fB\-lp\ \fIdirectory\fR\fR...] [\fB\-i\ \fIinit\-file\fR\fR...] [\fB\-c\fR\ [\fB\-l\fR]\ \fIlisp\-file\fR\ [\fB\-o\fR\fB\ \fR\fB\fIoutput\-file\fR\fR]...] [\fB\-x\ \fIexpressions\fR\fR...] [\fB\fIlisp\-file\fR\fR\ [\fB\fIargument\fR\fR...]]
.SH "DESCRIPTION"
.PP
Invokes the
\fI\fBCommon Lisp\fR\fR\&[1]
interpreter and compiler\. When called without arguments, executes the
\fIread\-eval\-print loop\fR\&[2], in which expressions are in turn
\fI\fBREAD\fR\fR\&[3]
from the standard input,
\fI\fBEVAL\fR\fR\&[4]uated by the lisp interpreter, and their results are
\fI\fBPRINT\fR\fR\&[5]ed to the standard output\. Invoked with
\fB\-c\fR, compiles the specified lisp files to a platform\-independent
bytecode
which can be executed more efficiently\.
.SH "OPTIONS"
.PP
\fB\-h\fR
.br
\fB\-\-help\fR
.RS 4
Displays a help message on how to invoke
\fI\fBCLISP\fR\fR\&[6]\.
.RE
.PP
\fB\-\-version\fR
.RS 4
Displays the
\fI\fBCLISP\fR\fR\&[6]
version number, as given by the function
\fI\fBLISP\-IMPLEMENTATION\-VERSION\fR\fR\&[7], the value of the variable
\fI*FEATURES*\fR, as well some other information\.
.RE
.PP
\fB\-\-license\fR
.RS 4
Displays a summary of the licensing information, the
\fIGNU\fR\&[8]
\fIGPL\fR\&[9]\.
.RE
.PP
\fB\-help\-image\fR
.RS 4
Displays information about the
memory image
being invoked: whether is it suitable for scripting as well as the
\fB:DOCUMENTATION\fR
supplied to
\fBEXT:SAVEINITMEM\fR\.
.RE
.PP
\fB\-B\fR \fIlisp\-lib\-dir\fR
.RS 4
Specifies the installation directory\. This is the directory containing the linking sets and other data files\. This option is normally not necessary, because the installation directory is already built\-in into the
\fBclisp\fR
executable\. Directory
\fIlisp\-lib\-dir\fR
can be changed dynamically using the
\fISYMBOL\-MACRO\fR\&[10]
\fI*LIB\-DIRECTORY*\fR\.
.RE
.PP
\fB\-K\fR \fIlinking\-set\fR
.RS 4
Specifies the
linking set
to be run\. This is a directory (relative to the
\fIlisp\-lib\-dir\fR) containing at least a main executable (runtime) and an initial
memory image\. Possible values are
.PP
\fBbase\fR
.RS 4
the core
\fI\fBCLISP\fR\fR\&[6]
.RE
.PP
\fBfull\fR
.RS 4
core plus all the modules with which this installation was built, see
Section\ 32.2, \(lqExternal Modules\(rq\.
.RE
.sp
The default is
\fBbase\fR\.
.RE
.PP
\fB\-M\fR \fImem\-file\fR
.RS 4
Specifies the initial
memory image\. This must be a memory dump produced by the
\fBEXT:SAVEINITMEM\fR
function by this
\fBclisp\fR
runtime\.
It may have been compressed using
\fIGNU\fR\&[8]
\fI\fBgzip\fR\fR\&[11]\.
.RE
.PP
\fB\-m\fR \fImemory\-size\fR
.RS 4
Sets the amount of memory
\fI\fBCLISP\fR\fR\&[6]
tries to grab on startup\. The amount may be given as
.PP
\fIn\fR
.br
\fIn\fR\fBB\fR
.RS 4
measured in bytes
.RE
.PP
\fIn\fR
.br
\fIn\fR\fBW\fR
.RS 4
measured in machine words (4\(mu\fIn\fR
on 32\-bit platforms, 8\(mu\fIn\fR
on 64\-bit platforms)
.RE
.PP
\fIn\fR\fBK\fR
.br
\fIn\fR\fBKB\fR
.RS 4
measured in kilobytes
.RE
.PP
\fIn\fR\fBKW\fR
.RS 4
measured in kilowords
.RE
.PP
\fIn\fR\fBM\fR
.br
\fIn\fR\fBMB\fR
.RS 4
measured in megabytes
.RE
.PP
\fIn\fR\fBMW\fR
.RS 4
measured in megawords
.RE
.sp
The default is 3 megabytes\.
The argument is constrained above 100 KB\.
.sp
This version of
\fI\fBCLISP\fR\fR\&[6]
is not likely to actually use the entire
\fImemory\-size\fR
since
garbage\-collection will periodically reduce the amount of used memory\. It is therefore common to specify 10 MB even if only 2 MB are going to be used\.
.RE
.PP
\fB\-L\fR \fIlanguage\fR
.RS 4
Specifies the
language
\fI\fBCLISP\fR\fR\&[6]
uses to communicate with the user\. This may be
one of
\fBenglish\fR, \fBgerman\fR, \fBfrench\fR, \fBspanish\fR, \fBdutch\fR, \fBrussian\fR, \fBdanish\fR\. Other languages may be specified through the
\fIenvironment variable\fR\&[12]
\fBLANG\fR, provided the corresponding message catalog is installed\.
The language may be changed dynamically using the
\fISYMBOL\-MACRO\fR\&[10]
\fICUSTOM:*CURRENT\-LANGUAGE*\fR\.
.RE
.PP
\fB\-N\fR \fIlocale\-dir\fR
.RS 4
Specifies the base directory of locale files\.
\fI\fBCLISP\fR\fR\&[6]
will search its message catalogs in
\fI\fIlocale\-dir\fR\fR\fI/\fR\fI\fIlanguage\fR\fR\fI/LC_MESSAGES/clisp\.mo\fR\. This directory may be changed dynamically using the
\fISYMBOL\-MACRO\fR\&[10]
\fICUSTOM:*CURRENT\-LANGUAGE*\fR\.
.RE
.PP
\fB\-E\fR\fB\fIdomain\fR\fR\fB \fR\fB\fIencoding\fR\fR
.RS 4
Specifies the encoding used for the given domain, overriding the default which depends on the
\fIenvironment variable\fR\&[12]s
\fBLC_ALL\fR,
\fBLC_CTYPE\fR,
\fBLANG\fR\.
\fIdomain\fR
can be
.PP
\fBfile\fR
.RS 4
affecting
\fICUSTOM:*DEFAULT\-FILE\-ENCODING*\fR
.RE
.PP
\fBpathname\fR
.RS 4
affecting
\fICUSTOM:*PATHNAME\-ENCODING*\fR
.RE
.PP
\fBterminal\fR
.RS 4
affecting
\fICUSTOM:*TERMINAL\-ENCODING*\fR
.RE
.PP
\fBforeign\fR
.RS 4
affecting
\fICUSTOM:*FOREIGN\-ENCODING*\fR
.RE
.PP
\fBmisc\fR
.RS 4
affecting
\fICUSTOM:*MISC\-ENCODING*\fR
.RE
.PP
\fIblank\fR
.RS 4
affecting all of the above\.
.RE
.sp
.sp
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
Warning
Note that the values of these
\fISYMBOL\-MACRO\fR\&[10]s that have been saved in a
memory image
are ignored: these
\fISYMBOL\-MACRO\fR\&[10]s are reset based on the OS environment
\fBafter\fR
the
memory image
is loaded\. You have to use the
RC file,
\fICUSTOM:*INIT\-HOOKS*\fR
or
init function
to set them on startup, but it is best to set the aforementioned
\fIenvironment variable\fR\&[12]s appropriately for consistency with other programs\. See
Section\ 31.1, \(lqCustomizing CLISP Process Initialization and Termination\(rq\.
.RE
.PP
\fB\-q\fR
.br
\fB\-\-quiet\fR
.br
\fB\-\-silent\fR
.br
\fB\-v\fR
.br
\fB\-\-verbose\fR
.RS 4
Change verbosity level: by default,
\fI\fBCLISP\fR\fR\&[6]
displays a banner at startup and a good\-bye message when quitting, and initializes
\fI\fI*LOAD\-VERBOSE*\fR\fR\&[13]
and
\fI\fI*COMPILE\-VERBOSE*\fR\fR\&[14]
to
\fI\fBT\fR\fR\&[15], and
\fI\fI*LOAD\-PRINT*\fR\fR\&[13]
and
\fI\fI*COMPILE\-PRINT*\fR\fR\&[14]
to
\fI\fBNIL\fR\fR\&[16], as per
[ANSI CL standard]\. The first
\fB\-q\fR
removes the banner and the good\-bye message, the second sets variables
\fI\fI*LOAD\-VERBOSE*\fR\fR\&[13],
\fI\fI*COMPILE\-VERBOSE*\fR\fR\&[14]
and
\fICUSTOM:*SAVEINITMEM\-VERBOSE*\fR
to
\fI\fBNIL\fR\fR\&[16]\. The first
\fB\-v\fR
sets variables
\fICUSTOM:*REPORT\-ERROR\-PRINT\-BACKTRACE*\fR,
\fI\fI*LOAD\-PRINT*\fR\fR\&[13]
and
\fI\fI*COMPILE\-PRINT*\fR\fR\&[14]
to
\fI\fBT\fR\fR\&[15], the second sets
\fICUSTOM:*LOAD\-ECHO*\fR
to
\fI\fBT\fR\fR\&[15]\. These settings affect the output produced by
\fB\-i\fR
and
\fB\-c\fR
options\. Note that these settings persist into the
\fIread\-eval\-print loop\fR\&[2]\. Repeated
\fB\-q\fR
and
\fB\-v\fR
cancel each other, e\.g\.,
\fB\-q \-q \-v \-v \-v\fR
is equivalent to
\fB\-v\fR\.
.RE
.PP
\fB\-on\-error\fR \fIaction\fR
.RS 4
Override (or force) the batch mode imposed by
\fB\-c\fR,
\fB\-x\fR, and
\fIlisp\-file\fR, depending on
\fIaction\fR:.PP
appease
.RS 4
\fIcontinuable\fR\&[17]
\fIERROR\fR\&[18]s are turned into
\fIWARNING\fR\&[19]s (with
\fBEXT:APPEASE\-CERRORS\fR) other
\fIERROR\fR\&[18]s are handled in the default way
.RE
.PP
debug
.RS 4
\fIERROR\fR\&[18]s
\fI\fBINVOKE\-DEBUGGER\fR\fR\&[20]
(the normal
\fIread\-eval\-print loop\fR\&[2]
behavior)
.RE
.PP
abort
.RS 4
\fIcontinuable\fR\&[17]
\fIERROR\fR\&[18]s are appeased, other
\fIERROR\fR\&[18]s are
\fI\fBABORT\fR\fR\&[21]ed with
\fBEXT:ABORT\-ON\-ERROR\fR
.RE
.PP
exit
.RS 4
\fIcontinuable\fR\&[17]
\fIERROR\fR\&[18]s are appeased, other
\fIERROR\fR\&[18]s terminate
\fI\fBCLISP\fR\fR\&[6]
with
\fBEXT:EXIT\-ON\-ERROR\fR
.RE
.sp
See also
\fBEXT:SET\-GLOBAL\-HANDLER\fR\.
.RE
.PP
\fB\-repl\fR
.RS 4
Start an interactive
\fIread\-eval\-print loop\fR\&[2]
after processing the
\fB\-c\fR,
\fB\-x\fR, and
\fIlisp\-file\fR
options and on any
\fIERROR\fR\&[18]
\fI\fBSIGNAL\fR\fR\&[22]ed during that processing\.
.RE
.PP
\fB\-w\fR
.RS 4
Wait for a keypress after program termination\.
.RE
.PP
\fB\-I\fR
.RS 4
Interact better with
\fIEmacs\fR\&[23]
(useful when running
\fI\fBCLISP\fR\fR\&[6]
under
\fIEmacs\fR\&[23]
using
\fISLIME\fR\&[24],
\fIILISP\fR\&[25]
et al)\. With this option,
\fI\fBCLISP\fR\fR\&[6]
interacts in a way that
\fIEmacs\fR\&[23]
can deal with:
.sp
.RS 4
\h'-04'\(bu\h'+03'unnecessary prompts are not suppressed\.
.RE
.RE
.PP
\fB\-ansi\fR
.RS 4
Comply with the
[ANSI CL standard]
specification even where
\fI\fBCLISP\fR\fR\&[6]
has been traditionally different by setting the
\fISYMBOL\-MACRO\fR\&[10]
\fICUSTOM:*ANSI*\fR
to
\fI\fBT\fR\fR\&[15]\.
.RE
.PP
\fB\-traditional\fR
.RS 4
Traditional: reverses the residual effects of
\fB\-ansi\fR
in the saved
memory image\.
.RE
.PP
\fB\-modern\fR
.RS 4
Provides a modern view of symbols: at startup the
\fI\fI*PACKAGE*\fR\fR\&[27]
variable will be set to the
\(lqCS\-COMMON\-LISP\-USER\(rq
package, and the
\fI\fI*PRINT\-CASE*\fR\fR\&[28]
will be set to
\fB:DOWNCASE\fR\. This has the effect that symbol lookup is case\-sensitive (except for keywords and old\-style packages) and that keywords and uninterned symbols are printed with lower\-case preferrence\. See
Section\ 11.5, \(lqPackage Case-Sensitivity\(rq\.
.RE
.PP
\fB\-p\fR \fIpackage\fR
.RS 4
At startup the value of the variable
\fI\fI*PACKAGE*\fR\fR\&[27]
will be set to the package named
\fIpackage\fR\. The default is the value of
\fI\fI*PACKAGE*\fR\fR\&[27]
when the image was
saved, normally
\fI\(lqCOMMON\-LISP\-USER\(rq\fR\&[29]\.
.RE
.PP
\fB\-C\fR
.RS 4
Compile when loading: at startup the value of the variable
\fICUSTOM:*LOAD\-COMPILING*\fR
will be set to
\fI\fBT\fR\fR\&[15]\. Code being
\fI\fBLOAD\fR\fR\&[30]ed will then be
\fI\fBCOMPILE\fR\fR\&[31]d on the fly\. This results in slower loading, but faster execution\.
.RE
.PP
\fB\-norc\fR
.RS 4
Normally
\fI\fBCLISP\fR\fR\&[6]
loads the user
\fI\(lqrun control\(rq (RC)\fR\&[32]
file on startup (this happens
\fBafter\fR
the
\fB\-C\fR
option is processed)\. The file loaded is
\fI\.clisprc\.lisp\fR
or
\fI\.clisprc\.fas\fR
in the home directory
\fI\fBUSER\-HOMEDIR\-PATHNAME\fR\fR\&[33], whichever is newer\. This option,
\fB\-norc\fR, prevents loading of the
RC file\.
.RE
.PP
\fB\-lp\fR \fIdirectory\fR
.RS 4
Specifies directories to be added to
\fICUSTOM:*LOAD\-PATHS*\fR
at startup\. This is done
\fBafter\fR
loading the
RC file
(so that it does not override the command\-line option) but
\fBbefore\fR
loading the init\-files specified by the
\fB\-i\fR
options (so that the init\-files will be searched for in the specified directories)\. Several
\fB\-lp\fR
options can be given; all the specified directories will be added\.
.RE
.PP
\fB\-i\fR \fIinit\-file\fR
.RS 4
Specifies initialization files to be
\fI\fBLOAD\fR\fR\&[30]ed at startup\. These should be lisp files (source or compiled)\. Several
\fB\-i\fR
options can be given; all the specified files will be loaded in order\.
.RE
.PP
\fB\-c\fR \fIlisp\-file\fR
.RS 4
Compiles the specified
\fIlisp\-file\fRs to bytecode (\fI*\.fas\fR)\. The compiled files can then be
\fI\fBLOAD\fR\fR\&[30]ed instead of the sources to gain efficiency\.
.RE
.PP
\fB\-o\fR \fIoutputfile\fR
.RS 4
Specifies the output file or directory for the compilation of the last specified
\fIlisp\-file\fR\.
.RE
.PP
\fB\-l\fR
.RS 4
Produce a bytecode
\fI\fBDISASSEMBLE\fR\fR\&[34]
listing (\fI*\.lis\fR) of the files being compiled\. Useful only for debugging\. See
Section\ 24.1, \(lqFunction COMPILE-FILE\(rq
for details\.
.RE
.PP
\fB\-x\fR \fIexpressions\fR
.RS 4
Executes a series of arbitrary expressions instead of a
\fIread\-eval\-print loop\fR\&[2]\. The values of the expressions will be output to
\fI\fI*STANDARD\-OUTPUT*\fR\fR\&[35]\. Due to the argument processing done by the shell, the
\fIexpressions\fR
must be enclosed in double quotes, and double quotes and backslashes must be escaped with backslashes\.
.RE
.PP
\fIlisp\-file\fR [ \fIargument\fR \.\.\. ]
.RS 4
Loads and executes a
\fIlisp\-file\fR, as described in
Script execution\. There will be no
\fIread\-eval\-print loop\fR\&[2]\. Before
\fIlisp\-file\fR
is loaded, the variable
\fIEXT:*ARGS*\fR
will be bound to a list of strings, representing the
\fIargument\fRs\.
The first line of

RUN_TIME/tmp/clisp-2.45/src/clisp.c
/*
 * Driver program for CLISP.
 *
 * Needed so that we can write #!/usr/local/bin/clisp as the first line
 * of a Lisp script. Normally, only real executables and no shell scripts
 * can be mentioned after #!.
 *
 * Since we are at it, this driver program also implements the "-K" option.
 * All other options are passed to the main program.
 *
 * Bruno Haible 31.3.1997
 * Sam Steingold 1998-2008
 */

/*
 * Macros passed during compilation:
 * LISPLIBDIR  string containing the directory with lisp.run and lispinit.mem
 * LOCALEDIR   string containing the locale directory
 */


/* Declare strlen(), strcpy(), strcat(). */
# include <string.h>
/* Declare stderr. */
# include <stdio.h>


# include <stdlib.h>            /* getenv, abort */
/* needed for execname.c to work */
# include <sys/types.h>         /* stat */
# include <sys/stat.h>          /* stat */
# include <unistd.h>            /* access */
# include <errno.h>             /* ENOMEM, ENOENT, errno */
# include <sys/param.h>         /* MAXPATHLEN */
int find_executable (const char * program_name);
/*
 * Finding the full path of the executable.
 * Bruno Haible 20.12.1994
 * Sam Steingold 2004-2006, 2008
 */

/* This assumes that the executable is not removed or renamed while
   running. */

/* file name of the executable */
static char* executable_name = NULL;
#define default_executable_name  "lisp.run"

/* file descriptor of the executable
 (Only used to verify that we find the correct executable.) */
static int executable_fd = -1;

/* maybe_executable(pathname)
 checks whether a given pathname may belong to the executable. */
static int maybe_executable (const char * filename) {
  struct stat statexe;
  struct stat statfile;
  if (access(filename,R_OK|X_OK) < 0)
    return 0/*false*/;
  if (executable_fd < 0)
    return 1/*true*/;
  /* If we already have an executable_fd, check that filename points to
   the same inode. */
  if (fstat(executable_fd,&statexe) < 0)
    return 1/*true*/;
  if (stat(filename,&statfile) < 0)
    return 0/*false*/;
  if (statfile.st_dev
      && statfile.st_dev == statexe.st_dev
      && statfile.st_ino == statexe.st_ino)
    return 1/*true*/;
  return 0/*false*/;
}

/* return the executable name */
char *get_executable_name (void);
char *get_executable_name (void) { return executable_name; }

/* find_executable(program_name)
 is to be called immediately after the program starts,
 with program_name = argv[0],
 before any chdir() operation and before any setenv("PATH",...).
 It determines the full program path and opens a file descriptor to
 the executable, for later use.
 Return value is 0 if successful, -1 and errno set if not. */
int find_executable (const char * program_name) {
  /* Do not need to execute this more than once. */
  if (executable_name != NULL) return 0;
 #if defined(UNIX_LINUX) || defined(UNIX_CYGWIN32)
  { /* The executable is accessible as /proc/<pid>/exe. We try this first
   because it is safer: no race condition w.r.t. the file system. It may
   fail, however, if the user has not compiled /proc support into his
   kernel. */
    int fd = open("/proc/self/exe",O_RDONLY,my_open_mask);
    if (fd >= 0)
      executable_fd = fd;
  }
 #endif
  { /* Now we guess the executable's full path. We assume the executable
   has been called via execlp() or execvp() with properly set up argv[0].
   The login(1) convention to add a '-' prefix to argv[0] is not supported. */
    const char * p;
    for (p = program_name; *p; p++)
      if (*p == '/')
        goto has_slash;
  }
  { /* exec searches paths without slashes in the directory list given
       by $PATH. */
    const char * path = getenv("PATH");
    if (!(path==NULL)) {
      const char * p;
      const char * p_next;
      for (p = path; *p; p = p_next) {
        const char * q;
        unsigned long p_len;
        for (q = p; *q; q++) { if (*q == ':') break; }
        p_len = q-p; p_next = (*q=='\0' ? q : q+1);
        { /* We have a path item at p, of length p_len.
             Now concatenate the path item and program_name. */
          char * concat_name =
            (char*) malloc(p_len + strlen(program_name) + 2);
          if (concat_name == NULL) { errno = ENOMEM; goto notfound; }
          if (p_len == 0) {
            /* empty PATH element designates the current directory */
            strcpy(concat_name,program_name);
          } else {
            memcpy(concat_name, p, p_len);
            concat_name[p_len] = '/';
            strcpy(concat_name+p_len+1, program_name);
          }
          if (maybe_executable(concat_name)) {
            /* Assume we have found the executable */
            program_name = concat_name; goto resolve;
          }
          free(concat_name);
        }
      }
    }
    /* Not found in the PATH, assume the current directory. */
  }
 has_slash:
  /* exec treats paths containing slashes as relative to the current
     directory */
  if (maybe_executable(program_name)) {
   resolve:
    /* resolve program_name: */
#  if !defined(MAXPATHLEN)
#   define MAXPATHLEN 1024      /* see unix.d */
#  endif
    executable_name = (char*) malloc(MAXPATHLEN);
    if (executable_name == NULL) { errno = ENOMEM; goto notfound; }
    if (realpath(program_name,executable_name) == NULL) {
      free(executable_name); goto notfound;
    }
    return 0;
  }
  errno = ENOENT;
 notfound:
  executable_name = (char*)default_executable_name; return -1;
}



int main (int argc, char* argv[])
{
  char* lisplibdir;
  char* localedir;
  char* argv_lisplibdir = NULL;
  char* argv_linkingset = "base";
  char* argv_memfile = NULL;
  char* argv_localedir = NULL;
  char* program_name;
  /*
   * To determine whether -K was given, we go through the options.
   * Because when "clisp foo.lisp -K" is invoked, the "-K" is an argument
   * for foo.lisp, not for clisp. Therefore we have to stop when we encounter
   * the first non-option, non-option-argument. Unfortunately, we have to
   * know about all of clisp's options.
   */
  program_name = argv[0];
  /*
   * The program_name may be absolute or relative if "clisp" is called
   * directly. (For example, "sh /usr/local/bin/clisp ..." will make it
   * absolute, "time clisp ..." will make it relative.)
   * If "clisp" is used as a script interpreter, program_name will be
   * - the full absolute pathname, on SunOS 4, Solaris, HP-UX, IRIX,
   * - only the basename, on Linux, AIX, OSF/1.
   * It follows that we cannot tell whether we have been called as
   * script interpreter or directly.
   */
# if ENABLE_RELOCATABLE
  /* Put this executable's absolute path into executable_name. */
  if (find_executable(program_name) < 0) {
    fprintf(stderr,"%s: cannot figure out the absolute executable path",
            program_name);
    return 1;
  }
  /* Figure out lisplibdir and localedir. */
  {
    unsigned int libdir_len;
    const char *p;
    char *mem;
    /* The libdir is determined as `dirname $executable_name`. */
    for (p = executable_name + strlen(executable_name);; p--) {
      if (p == executable_name) abort();
      if (*p == '/') break;
    }
    libdir_len = p - executable_name;
    mem = (char*)malloc((libdir_len+1)+(libdir_len+7+1));
    if (mem == NULL) goto oom;
    lisplibdir = mem;
    localedir = mem + (libdir_len+1);
    /* Compute lisplibdir from it. */
    memcpy(lisplibdir,executable_name,libdir_len);
    lisplibdir[libdir_len] = '\0';
    /* Compute localedir from it. */
    memcpy(localedir,executable_name,libdir_len);
    localedir[libdir_len] = *p; /* directory separator */
    memcpy(localedir+libdir_len+1,"locale",6);
    localedir[libdir_len+7] = '\0';
  }
# else
  lisplibdir = LISPLIBDIR;
  localedir = LOCALEDIR;
# endif
  /*
   * Script execution on Unix is implemented like this:
   * - The basename/fullname of the interpreter is put into argv[0].
   * - (Optional - only if at least one interpreter-arg is present.) Next
   *   comes the tail of the "#!..." line. On SunOS 4, Linux, IRIX, AIX,
   *   OSF/1: with leading whitespace stripped, but whitespace inside it
   *   untouched (!). On Solaris, HP-UX: with leading whitespace stripped,
   *   and cut off at the next whitespace character (!!).
   * - Next comes the filename of the script.
   * - Then all the arguments of the script.
   * We therefore need to split argv[1] into pieces. We shouldn't split
   * "-x" arguments into pieces. However, fortunately, the "-x" argument
   * cannot appear as argv[1] (because "-x" must precede it), and it
   * cannot appear within the "#!..." line (because "#!" and "-x" are
   * mutually exclusive).
   * As a workaround against the Solaris/HP-UX problem, we split not
   * only at normal spaces, but also at hard spaces.
   * See <impnotes.html#quickstart>.
   */
  if (argc > 1) {
    int wordcount = 0; /* number of pieces in argv[1] */
    { char* arg = argv[1];
      int inword = 0;
      while (*arg != '\0') {
        int spacep = (*arg == '\t' || *arg == ' ' || *arg == (char)0xA0);
        if (!inword && !spacep) wordcount++;
        inword = !spacep;
        arg++;
      }
    }
    {int old_argc = argc;
     char** old_argv = argv;
     int new_argc = argc + wordcount - 1;
     char** new_argv = (char**) malloc((new_argc+1)*sizeof(char*));
     if (!new_argv) goto oom;
     argc = new_argc;
     argv = new_argv;
     /* Copy argv[0] unmodified. */
     *new_argv++ = *old_argv++;
     { /* Split argv[1] into pieces. */
       char* arg = *old_argv++;
       int inword = 0;
       while (*arg != '\0') {
         int spacep = (*arg == '\t' || *arg == ' ' || *arg == (char)0xA0);
         if (!inword) {
           if (!spacep) { *new_argv++ = arg; }
         } else {
           if (spacep) { *arg = '\0'; }
         }
         inword = !spacep;
         arg++;
       }
     }
     { /* Copy argv[2..argc-1] unmodified. */
       int i;
       for (i = old_argc-2; i > 0; i--) { *new_argv++ = *old_argv++; }
     }
     *new_argv = NULL;
    }
  }
  /*
   * Done with script interpreter argument handling.
   */
  { char** argptr = &argv[1];
    char** argptr_limit = &argv[argc];
    enum { for_exec, for_init, for_compile } argv_for = for_exec;
    while (argptr < argptr_limit) {
      char* arg = *argptr++;
      if ((arg[0] == '-') && !(arg[1] == '\0')) {
        switch (arg[1]) {
#        define OPTION_ARG  \
          if (arg[2] == '\0') \
            { if (argptr < argptr_limit) arg = *argptr++; else goto usage; } \
          else { arg = &arg[2]; }
          /* Options to which we have to pay attention. */
          case 'B':
            OPTION_ARG;
            lisplibdir = argv_lisplibdir = arg;
            break;
          case 'K':
            OPTION_ARG;
            argv_linkingset = arg;
            break;
          case 'M':
            OPTION_ARG;
            argv_memfile = arg;
            break;
          case 'N':
            OPTION_ARG;
            argv_localedir = arg;
            break;
          /* Skippable options without arguments. */
          case 'h':
          case 'd':
          case 'q':
          case 'I':
          case 'C':
          case 'l':
          case 'a':
          case 'w':
          case 'n': /* -norc */
          case 'r': /* -repl */
          case 'v':
            break;
          case '-':
            if (arg[2] == '\0') goto done_options; /* --: end of arguments */
            else break;   /* GNU-style long options --help, --version */
          /* Skippable options with arguments. */
          case 'm':
          case 't':
          case 'L':
          case 'o':
          case 'p':
          case 'x':
            OPTION_ARG;
            break;
          case 'E':
            if (argptr < argptr_limit) argptr++; else goto usage;
            break;
          case 'i':
            if (arg[2] == '\0') argv_for = for_init;
            break;
          case 'c':
            argv_for = for_compile;
            break;
          default:
            goto usage;
        }
      } else {
        switch (argv_for) {
          case for_init:
          case for_compile:
            break;
          case for_exec:
            /* All the remaining options are for the Lisp program. */
            goto done_options;
        }
      }
    }
   done_options: ;
  }
  { char* linkingsetdir;
    char* executable;
    char** new_argv;
    /* Compute linking set. */
    if (argv_linkingset == NULL || strlen(argv_linkingset)==0) {
      linkingsetdir = lisplibdir;
    } else if (argv_linkingset[0]=='/') {
      linkingsetdir = argv_linkingset;
    } else {
      linkingsetdir =
        (char*)malloc(strlen(lisplibdir)+1+strlen(argv_linkingset)+1);
      if (!linkingsetdir) goto oom;
      strcpy(linkingsetdir, lisplibdir);
      strcat(linkingsetdir, "/");
      strcat(linkingsetdir, argv_linkingset);
    }
    { /* Compute executable's name. */
      char* execname = "lisp.run";
      executable = (char*)malloc(strlen(linkingsetdir)+1+strlen(execname)+1);
      if (!executable) goto oom;
      strcpy(executable, linkingsetdir);
      strcat(executable, "/");
      strcat(executable, execname);
    }
    /* Compute new arguments. */
    new_argv = (char**)malloc((argc+6+1)*sizeof(char*));
    if (!new_argv) goto oom;
    new_argv[0] = executable;
    { char** argptr = &argv[1];
      char** argptr_limit = &argv[argc];
      char** new_argptr = &new_argv[1];
      if (!argv_lisplibdir) {
        *new_argptr++ = "-B";
        *new_argptr++ = lisplibdir;
      }
      if (!argv_memfile) {
        char* filename = "lispinit.mem";
        argv_memfile =
          (char*)malloc(strlen(linkingsetdir)+1+strlen(filename)+1);
        if (!argv_memfile) goto oom;
        strcpy(argv_memfile, linkingsetdir);
        strcat(argv_memfile, "/");
        strcat(argv_memfile, filename);
        *new_argptr++ = "-M";
        *new_argptr++ = argv_memfile;
      }
      if (!argv_localedir) {
        *new_argptr++ = "-N";
        *new_argptr++ = localedir;
      }
      while (argptr < argptr_limit) { *new_argptr++ = *argptr++; }
      *new_argptr = NULL;
    }
    /* Launch the executable. */
    execv(executable,new_argv);
    { /* execv() returns only if there was an error. */
      int saved_errno = errno;
      fprintf(stderr,"%s: ",program_name);
      errno = saved_errno; perror(executable);
    }
    return 1;
  }
  usage: {
    fprintf(stderr,"%s: invalid command-line option, try `%s --help'\n",
            program_name,program_name);
    return 1;
  }
  oom: {
    fprintf(stderr,"%s: out of memory\n",program_name);
    return 1;
  }
}
